x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportGetMlAutoscalingStats.java

```diff
@@ -61,26 +58,28 @@ public TransportGetMlAutoscalingStats(
         );
         this.mlMemoryTracker = mlMemoryTracker;
         this.settings = settings;
-        this.timeoutExecutor = threadPool.generic();
     }
 
     @Override
     protected void masterOperation(Task task, Request request, ClusterState state, ActionListener<Response> listener) {
 
-        if (mlMemoryTracker.isRecentlyRefreshed()) {
-            MlAutoscalingResourceTracker.getMlAutoscalingStats(
-                state,
-                clusterService.getClusterSettings(),
-                mlMemoryTracker,
-                settings,
-                ActionListener.wrap(autoscalingResources -> listener.onResponse(new Response(autoscalingResources)), listener::onFailure)
-            );
-        } else {
-            // Recent memory statistics aren't available at the moment, trigger a refresh and return a no-scale.
-            // (If a refresh is already in progress, this won't trigger a new one.)
+        if (mlMemoryTracker.isRecentlyRefreshed() == false) {
+            // Recent memory statistics aren't available at the moment, trigger a refresh in the background.
+            // (If a refresh is already in progress, this won't trigger a new one.) We still proceed to return a
+            // scaling response in this case. This API gets called every 5 seconds, so the memory info is likely only
```

Comment:
Unrelated to your changes but if this is called every 5 seconds is it possible that this block would get executed every 5 seconds meaning a refresh every 5 seconds? Maybe a refresh every 5 seconds isn't that bad.

Reply:
Yes, we can ask for a refresh every 5 seconds. But the actual refresh code in the memory tracker itself doesn't do another refresh if there's already one in progress. It's here: https://github.com/elastic/elasticsearch/blob/995b4d3c69e7ab06c994da67b2692615aa69a15a/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/process/MlMemoryTracker.java#L411